### ISSUE TITLE
[PRE_COMMANDS] Bug / Regression or Expected behavior

### ISSUE THREAD

**Issue by llaville** [megalinter/megalinter#819]
While I'm on final step to solve issue #725, on applying solution described behind https://github.com/nvuillam/mega-linter/issues/725#issuecomment-937593356, I found a possible regression about pre-commands feature.

Especially if I've well understood what I read on v4.45.0 

![tag_v4 45 0](https://user-images.githubusercontent.com/364342/136391766-69ba825d-2bc4-4020-9d1b-fea5e76986ef.png)
 
Can I declare a default pre-command at linter level like that ? 

![linter_desc_pre_commands](https://user-images.githubusercontent.com/364342/136392060-527e6507-224a-4eb3-a41a-ae8e73b462e9.png)

If answer is YES, than there is a regression in v4.46 !

Here is what I got 

![phpstan_pre_command](https://user-images.githubusercontent.com/364342/136392404-7f55e4a9-980b-4e49-bb5b-2563a0681372.png)

And, what I expected to have 

![phpstan_command](https://user-images.githubusercontent.com/364342/136392490-22040dfd-70c9-499b-91f1-f8f6241d70b3.png)

With final result (poor performance compare to `list_of_files` mode : see #817)

![phpstan_summary](https://user-images.githubusercontent.com/364342/136392843-3de7dcb4-08d1-423f-a53f-bf105edfb900.png)

**Comment by llaville** [comment]
@nvuillam Reminder to get a response to clarify this feature before github-actions bot close this report

**Comment by nvuillam** [comment]
Sorry this issue seems quite complicated
I think that pre-commands should be used mostly to install libter plugins or change default behaviors, dor example by setting env vars

Why do you need to copy files ?

**Comment by llaville** [comment]
I really want a clear situation about architecture expected PRE|POST COMMANDS behavior.
I thought, I could do any action before/after linter was run, but it seems its not TRUE. Example with linter config files that once detected could not be changed later (in my case copy config file from `.github/linters` to root project).

**Comment by llaville** [comment]
When its too complicated to explain to a child, the system should be too complex, and must be simplified.

**Comment by nvuillam** [comment]
I still don't get why you want to copy files in PRE_COMMANDS ^^

**Comment by llaville** [comment]
No more needed to use the PRE_COMMANDS in context of #725 (found origin of problem). 
PHPStan linter has a special behavior with config file and data source scanned, that Mega-Linter v5 does not follow except if we know the tip ;-) I'm preparing a full summary report.
BTW, I recommand to open a Best Practices section in documention or open the GitHub Discussion in this repo to list a FAQ.

**Comment by llaville** [comment]
Origin of issue with PHPStan come from https://github.com/megalinter/megalinter/blob/v5.2.0/megalinter/Linter.py#L632-L636

Don't know why such code; is it really necessary ? At lease PHPStan to work correctly requires to run in project folder. Currently we have 
```
[phpstan] command: ['phpstan', 'analyse', '--no-progress', '--no-ansi', '--memory-limit', '1G', '-c', '/action/lib/.automation/phpstan.neon.dist', '/tmp/lint/tests/TestCase.php', '/tmp/lint/tests/TestHandler.php']
[phpstan] CWD: /
```
`CWD = /`  => outside project folder context, this is why PHPStan do not work well !

**TIP**  Run docker locally with `-w /tmp/lint` locally  solve the problem

i.e:
```
sudo rm -rf report/ && time docker run -v $(pwd):/tmp/lint -e LOG_LEVEL=DEBUG -e ENABLE=PHP -w /tmp/lint megalinter/megalinter-php:v5```

**Comment by llaville** [comment]
PRE-COMMANDs feature in state is quite difficult to solve problem that a hook system should be able to do !
I'll then close this report.
